V2/dev change get params return type

c-w-feldmann opened the pull request:
## Description
We wrote a sklearn adapter for the chemprop class which strongly relies on the `get_params` method. 
Since the signature here differs to the sklearn method, this regularly causes problems and we wanted to ask if it is possible to adapt the function here.  
See also the related [issue](https://github.com/chemprop/chemprop/issues/539) 

All tests are passed and code was standardized with black.

davidegraff commented on chemprop/nn/loss.py:
```diff
@@ -149,8 +151,23 @@ def forward(self, preds: Tensor, targets: Tensor, *args) -> Tensor:
-    def get_params(self) -> list[tuple[str, float]]:
-        return [("v_kl", self.v_kl), ("eps", self.eps)]
+    def get_params(self, deep: bool = True) -> dict[str, float]:  # pylint: disable=unused-argument
+        """Get the parameters of the class.
+
+        Overrides the `ReprMixin.get_params` method to return the `v_kl` and `eps` parameter.
+
+        Parameters
+        ----------
+        deep: bool, default=True
+            No effect on this method.
+
+        Returns
+        -------
+        dict[str, float]
+            The parameters of the class.
+        """
+        param_dict = {"v_kl": self.v_kl, "eps": self.eps}
+        return param_dict
```
The original docstring is still correct--there's no use in overriding it.
```suggestion
        return {"v_kl": self.v_kl, "eps": self.eps}
```

davidegraff replied:
The parent docstring reads:
```
        deep : bool, default=True
            If True, will return the parameters for this estimator and
            contained subobjects that are estimators.
```
This is a no-op for an object that contains no child objects (i.e., this one.) Because it's a no-op for this object **does not** mean it's no longer correct.

davidegraff replied:
>  A broader question: Why is `get_params()` overwritten for `EvidentialLoss`, `DirichletMixin`, and `_ThresholdMixin` anyway?
> If they produce essentially the same output, does overwriting it have a different purpose?

I believe this is a vestige of the original implementation, where I could not rely on the default implementation to work in some cases, so I had to override it here to ensure that the correct representation was returned. It now needs to be overridden in cases where the input parameter name _does not_ match the internal attribute name, i.e.:
```python
class Foo:
    def __init__(self, a: int = 5):
        self.b = a
```
In both the current and proposed implementations, we have the following:
```python
>>> Foo()
Foo()
>>> Foo(10)
Foo()
```
even though it should read:
```python
>>> Foo(10)
Foo(a=10)
```
Fixing this however would require some changes and that brings this PR out-of-sync with the `sklearn` implementation (because ours would be a general solution to the representation problem). I don't personally care about this issue TBH as long as there's a note that the class only works when parameter names match their corresponding attribute names.

davidegraff commented on chemprop/nn/loss.py:
```diff
@@ -267,8 +284,24 @@ def forward(self, preds, targets, *args) -> Tensor:
 
-    def get_params(self) -> list[tuple[str, float]]:
-        return [("v_kl", self.v_kl)]
+    def get_params(
+            self, deep: bool = True  # pylint: disable=unused-argument
+    ) -> dict[str, float]:
+        """Get the parameters of the class.
+
+        Overrides the `ReprMixin.get_params` method to return the `v_kl` parameter.
+
+        Parameters
+        ----------
+        deep: bool, default=True
+            No effect on this method.
+
+        Returns
+        -------
+        dict[str, float]
+            The parameters of the class.
+        """
```
```suggestion
    def get_params(
            self, deep: bool = True  # pylint: disable=unused-argument
    ) -> dict[str, float]:
```

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -1,20 +1,57 @@
+        non_default_params = ", ".join(param_repr_list)
+        return f"{self.__class__.__name__}({non_default_params})"
+
+    def _get_default_params(self) -> dict[str, Any]:
+        """Get the default parameters for initializing the class."""
+        sig = inspect.signature(self.__class__)
+        return {k: v.default for k, v in sig.parameters.items()}
```
```suggestion
    @classmethod
    def _get_default_params(cls) -> dict[str, Any]:
        """Get the default parameters for initializing the class."""
        sig = inspect.signature(cls)
        return {k: v.default for k, v in sig.parameters.items()}
```

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -1,20 +1,57 @@
+"""Mixin classes for the project."""
+
```
```suggestion
```

davidegraff commented on chemprop/nn/loss.py:
```diff
@@ -1,3 +1,5 @@
+"""Loss functions for use in training models."""
+
```
I'd rather stick to the general convention of no module-level docstrings for now. This can change in a future PR though.
```suggestion
```

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -1,20 +1,57 @@
+            if param_value != param_default:
+                param_repr_list.append(f"{param_name}={repr(param_value)}")
+
+        non_default_params = ", ".join(param_repr_list)
+        return f"{self.__class__.__name__}({non_default_params})"
```
```suggestion
        return f"{self.__class__.__name__}({','.join(param_repr_list)})"
```

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -1,20 +1,57 @@
+        sig = inspect.signature(self.__class__)
+        return {k: v.default for k, v in sig.parameters.items()}
+
+    def _get_parm_names(self) -> list[str]:
+        """Get the default parameters for initializing the class."""
```
I don't think this docstring is correct.
```suggestion
    @classmethod
    def _get_param_names(cls) -> list[str]:
        """Get the default parameters for initializing the class."""
```

davidegraff replied:
Also, why even define this method at all? It's just syntactic sugar for `self._get_default_params().keys()`. It's not  clear to me what advantage abstracting that call into a method confers.

KnathanM replied:
It looks like c-w-feldmann was following the style of scikit-learn, which has `_get_param_names()` as a separate method for deprecation and error checking (best I can tell). We don't necessarily need those for Chemprop and could simplify this by just calling `self._get_default_params().keys()` as you suggested.

c-w-feldmann replied:
The docstring is indeed wrong. I'll fix it.  
I used sorted, as the `keys()` method returns a `dict_keys` object and not a `list`.
(Using `list` would equally work, but I had some years ago problems with reproducing results exactly, where scripts used `sets`. So maybe I am just a bit paranoid.)
This is just nitpicking, so I am also fine with returning `self._get_default_params().keys()`

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -1,20 +1,57 @@
+        """Get the current parameters of the class, required to initialize the class.
 
-        argspec = ", ".join(f"{k}={repr(v)}" for k, v in items)
+        This method mimics the behavior of `sklearn.base.BaseEstimator.get_params`.
```
```suggestion
        This method mimics the behavior of :method:`sklearn.base.BaseEstimator.get_params`.
```

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -1,20 +1,57 @@
+
+    def _get_parm_names(self) -> list[str]:
+        """Get the default parameters for initializing the class."""
+        return sorted(self._get_default_params().keys())
```
why `sorted`?

KnathanM replied:
scikit-learn does sorted, maybe to be easier to understand if there are lots of keys? I don't have a preference either way.

davidegraff replied:
IMO it's clearer without `sorted`--then the list of param names follows the initializer signature, i.e., the source.

c-w-feldmann replied:
Copied from the [answer above](https://github.com/chemprop/chemprop/pull/649#discussion_r1492516015):  
I used sorted, as the `keys()` method returns a `dict_keys` object and not a `list`.
(Using list would equally work, but I had some years ago problems with reproducing results exactly, where scripts used sets. So maybe I am just a bit paranoid.)

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -1,20 +1,57 @@
+        dict[str, Any]
+            Parameter names mapped to their values.
+        """
+        out: dict[str, Any] = {}
+        for param_name in self._get_parm_names():
+            value = getattr(self, param_name)
+            if deep and hasattr(value, "get_params") and not isinstance(value, type):
+                deep_items = value.get_params().items()
+                out.update((param_name + "__" + k, val) for k, val in deep_items)
+            out[param_name] = value
+        return out
```
```suggestion
        name_to_value = {}
        for name in self._get_default_params.keys():
            value = getattr(self, param_name)
            name_to_value[name] = value
            if deep and hasattr(value, "get_params") and not isinstance(value, type):
                _params = value.get_params()
                name_to_value.update((f"{name}__{k}", v) for k, v in _params.items())
            
        return name_to_value
```

KnathanM replied:
Looks like this function was copied straight from scikit-learn. Would probably be good to edit it to fit the style of Chemprop more as you have done. Could also be good to leave it identical to scikit-learn if that has worked for scikit-learn.

davidegraff reviewed (changes requested):
FWIW I don't think the `ReprMixin` class adds much value to the repo. I'd prefer to just make `Metric` and `LossFunction` subclass `nn.Module` and get rid of this class entirely.

KnathanM reviewed (commented):
Thank you @c-w-feldmann for working on this. And thank you @davidegraff for your review. If you have any questions about our comments @c-w-feldmann, please reply to them so we can clarify. I hope this helps get a step closer to having Chemprop working well for you.

c-w-feldmann commented:
Except for the discussion about the implementation of `_get_params` ([discussion](https://github.com/chemprop/chemprop/pull/649#discussion_r1491205952)), I think I addressed all comments.  
It was mentioned that `ReprMixin` might also be removed. If you want to, I could also address this.

davidegraff commented on chemprop/nn/loss.py:
```diff
@@ -294,8 +296,22 @@ def forward(self, preds: Tensor, targets: Tensor, mask: Tensor, *args) -> Tensor
+    def get_params(
+        self, deep: bool = True  # pylint: disable=unused-argument
+    ) -> dict[str, float | None]:
+        """Get the parameters of the class.
+
+        Parameters
+        ----------
+        deep: bool, default=True
+            No effect on this method.
+
+        Returns
+        -------
+        dict[str, float | None]
+            The parameters of the class.
+        """
```
```suggestion
```

c-w-feldmann replied:
Removed the docstring

davidegraff commented:
@c-w-feldmann just FYI you can add all suggestions into a single batch commit rather than committing them one at a time (nicer to everyone's email!)

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -4,17 +4,47 @@
+        Parameters
+        ----------
+        deep : bool, default=True
+            If True, will return the parameters for this estimator and
+            contained subobjects that are estimators.
```
Is this correct?

KnathanM replied:
This is copied from [scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/base.py#L234). I see the function checks if the contained thing `hasattr(value, "get_params") and not isinstance(value, type)`. Perhaps those two conditions mean the contained thing is an estimator in scikit-learn. 

But in our case, `get_params()` is only implemented for the loss functions, which aren't estimators (based on my understanding of the scikit-learn [docs](https://scikit-learn.org/stable/developers/develop.html)).

davidegraff replied:
It's clear this is copied from sklearn because we have no concept of an `Estimator` in our code

KnathanM replied:
Then perhaps a more correct description would be "If True, will return the parameters for this object and all contained subobjects."

c-w-feldmann replied:
updated the docstring

davidegraff commented on chemprop/utils/mixins.py:
```diff
@@ -4,17 +4,47 @@
+        Returns
+        -------
+        dict[str, Any]
+            Parameter names mapped to their values.
```
can this be more descriptive? Something like:
```suggestion
            The parameter names and respective values used to initialize this class. That is, an identical initialization of this object can be achieved via:
            
            ```
            o: ReprMixin
            o_copy = type(o)(**o.get_params())
            ```
```

davidegraff replied:
though I don't think this is correct in the case of `deep=True`...

c-w-feldmann replied:
Updated the docstring

KnathanM commented:
@c-w-feldmann Can you remind me your specific use-case for wanting to change the behavior of `get_params()`? In scikit-learn, `get_params()` is used for standard API for estimators (models). In Chemprop, `get_params()` has the limited use case to programmatically reproduce loss functions (not models), as David showed in a comment `o_copy = type(o)(**o.get_params())`. Knowing your use case may help us know how much of the scikit-learn style functions to include.

c-w-feldmann commented:
@KnathanM

> Can you remind me your specific use-case for wanting to change the behavior of get_params()?  

We have a ml workflow which is expecting sklearn behaviour. For chemprop we wrote corresponding wrappers, however the loss function implements the `get_params` function as well, but has a different behavoir.
So when calling  `get_params` on the wrapped model,  `get_params` of its instance variables is called (if available), resulting for loss-objects in:
```python
Traceback (most recent call last):
  File "xyz123/test_chemprop/test_component_wrapper.py", line 11, in test_get_set_params
    params = binary_clf_ffn.get_params(deep=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "xyz456/lib/python3.11/site-packages/sklearn/base.py", line 246, in get_params
    deep_items = value.get_params().items()
                 ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'dict_items' object has no attribute 'items'
```
If you are interested in the wrapper themself I think we can share them with you, but I check again first with my team.

c-w-feldmann commented:
@davidegraff, @KnathanM I addressed your comments a while ago. Did you have time to have a look?

KnathanM commented:
Thank you @c-w-feldmann for pinging us again on this. 

I have tested the changes and get the same `repr(obj)` behavior for the loss functions, so I think this PR would be okay to merge in (probably squash merge). 

Alternatively, @davidegraff , why do some of the loss functions have their own `get_params()` method (for example EvidentialLoss)? If I comment them out I get the same `repr` behavior. Could we instead just change `ReprMixin` to not use `get_params() like this?
```
class ReprMixin:
    def __repr__(self) -> str:
        items = self.__dict__.items()

        if len(items) > 0:
            keys, values = zip(*items)
            sig = inspect.signature(self.__class__)
            defaults = [sig.parameters[k].default for k in keys]
            items = [(k, v) for k, v, d in zip(keys, values, defaults) if v != d]

        argspec = ", ".join(f"{k}={repr(v)}" for k, v in items)

        return f"{self.__class__.__name__}({argspec})"
```

@c-w-feldmann , If chemprop loss functions didn't have a `get_params()` method, would that help your sklearn wrappers to work correctly?

davidegraff commented:
I'm in favor of closing this PR and moving `nn.LossFunction` over to using [`extra_repr()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.extra_repr) rather than defining some custom `__repr__()` method for this class. This is the "official" way to provide extra information about an `nn.Module` rather than what I had initially written.

KnathanM commented:
> I'm in favor of closing this PR and moving `nn.LossFunction` over to using [`extra_repr()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.extra_repr) rather than defining some custom `__repr__()` method for this class. This is the "official" way to provide extra information about an `nn.Module` rather than what I had initially written.

Do you mean like this: #758 ?

davidegraff commented:
> > I'm in favor of closing this PR and moving `nn.LossFunction` over to using [`extra_repr()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.extra_repr) rather than defining some custom `__repr__()` method for this class. This is the "official" way to provide extra information about an `nn.Module` rather than what I had initially written.
> 
> Do you mean like this: #758 ?

Yeah that's on the right track

KnathanM commented:
@c-w-feldmann Do the changes in #758 allow your sklearn wrapper to work? The changes in that PR fit the style of Chemprop better than the changes here that are adapted from sklearn. But I also want to make sure you can use Chemprop in your workflow. We are hoping to finish a formal release of v2.0 stable in the middle of next week and could include changes with that.

c-w-feldmann commented:
> Do the changes in https://github.com/chemprop/chemprop/pull/758 allow your sklearn wrapper to work?

Yes that would work. Closing the PR.
